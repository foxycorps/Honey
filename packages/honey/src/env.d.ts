declare module 'slash' {
  export default function slash(path: string): string
}

declare module 'download' {
  export default function dwonload(url: string, path: string): Promise<void>
}

declare module 'github-slugger' {
  export default class Slugger {
    slug(data: string): string
  }
}
