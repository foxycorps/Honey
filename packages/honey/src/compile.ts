import { createProcessor, ProcessorOptions } from '@mdx-js/mdx'
import remarkGfm from 'remark-gfm'
import rehypePrettyCode from 'rehype-pretty-code'
import { remarkStaticImage } from './mdx-plugins/static-image'
import remarkHandler, { HeadingMeta } from './mdx-plugins/remark'
import { LoaderOptions } from './types'
import structurize from './mdx-plugins/structurize'
import { parseMeta, attachMeta } from './mdx-plugins/rehype-handler'

// @ts-ignore
import theme from './theme.json'

const createCompiler = (mdxOptions: ProcessorOptions) => {
  const compiler = createProcessor(mdxOptions)
  compiler.data('headingMeta', {
    hasH1: false,
    headings: []
  })
  return compiler
}

const rehypePrettyCodeOptions = {
  theme,
  // onVisitLine(node: any) {
  //   // Style a line node.
  //   Object.assign(node.style, {
  //   })
  // },
  onVisitHighlightedLine(node: any) {
    // Style a highlighted line node.
    if (!node.properties.className) {
      node.properties.className = []
    }
    node.properties.className.push('highlighted')
  },
  onVisitHighlightedWord(node: any) {
    // Style a highlighted word node.
    if (!node.properties.className) {
      node.properties.className = []
    }
    node.properties.className.push('highlighted')
  }
}

export async function compileMdx(
  source: string,
  mdxOptions: LoaderOptions['mdxOptions'] = {},
  honeyOptions: {
    unstable_staticImage: boolean
    unstable_flexsearch:
    | boolean
    | {
      codeblocks: boolean
    }
  } = {
      unstable_staticImage: false,
      unstable_flexsearch: false
    },
  resourcePath: string
) {
  let structurizedData = {}
  const compiler = createCompiler({
    jsx: true,
    providerImportSource: '@mdx-js/react',
    remarkPlugins: [
      ...(mdxOptions.remarkPlugins || []),
      remarkGfm,
      remarkHandler,
      ...(honeyOptions.unstable_staticImage ? [remarkStaticImage] : []),
      ...(honeyOptions.unstable_flexsearch
        ? [structurize(structurizedData, honeyOptions.unstable_flexsearch)]
        : [])
    ].filter(Boolean),
    // @ts-ignore
    rehypePlugins: [
      ...(mdxOptions.rehypePlugins || []),
      parseMeta,
      [rehypePrettyCode, rehypePrettyCodeOptions],
      attachMeta
    ].filter(Boolean)
  })
  try {
    const result = await compiler.process(source)
    return {
      result: String(result),
      ...(compiler.data('headingMeta') as HeadingMeta),
      structurizedData
    }
  } catch (err) {
    console.error(`\nError compiling ${resourcePath}`)
    console.error(`${err}\n`)
    throw err
  }
}
