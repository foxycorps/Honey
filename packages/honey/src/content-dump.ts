import fs from 'graceful-fs'
import path from 'path'

const { statSync, mkdirSync } = fs

const cacheDir = path.join(process.cwd(), '.next', 'cache')
const assetDir = path.join(process.cwd(), '.next', 'static', 'chunks')

const asset: { [locale: string]: any } = {}
const cached: Map<string, boolean> = new Map()

try {
  statSync(assetDir)
} catch (err) {
  mkdirSync(assetDir, { recursive: true })
}

let cacheDirExist = false
try {
  statSync(cacheDir)
  cacheDirExist = true
} catch (err) {
  mkdirSync(cacheDir, { recursive: true })
}

function initFromCache(filename: string) {
  if (!cached.has(filename)) {
    try {
      const content = fs.readFileSync(path.join(assetDir, filename)).toString()
      cached.set(filename, true)
      return JSON.parse(content)
    } catch (err) {
      cached.set(filename, false)
    }
  }
  return {}
}

export async function addPage({
  fileLocale,
  route,
  title,
  data,
  structurizedData
}: {
  fileLocale: string
  route: string
  title: string
  data: any
  structurizedData: any
}) {
  const dataFilename = `honey-data-${fileLocale}.json`

  if (!asset[fileLocale]) {
    asset[fileLocale] = initFromCache(dataFilename)
  }
  asset[fileLocale][route] = {
    title: title || data.title,
    data: structurizedData
  }

  // To prevent race conditions, we temporarily use the sync method to flush.
  // @TODO: introduce mutex lock, or only generate the asset when finishing the
  // entire build.
  const content = JSON.stringify(asset[fileLocale])
  fs.writeFileSync(path.join(assetDir, dataFilename), content)
  fs.writeFileSync(path.join(cacheDir, dataFilename), content)
}

// Copy cached data to the asset directory.
export async function restoreCache() {
  if (cacheDirExist) {
    try {
      statSync(assetDir)
    } catch (err) {
      mkdirSync(assetDir, { recursive: true })
    }

    const files = fs.readdirSync(cacheDir)
    for (const file of files) {
      if (file.startsWith('honey-data-')) {
        fs.copyFileSync(path.join(cacheDir, file), path.join(assetDir, file))
      }
    }
  }
}
