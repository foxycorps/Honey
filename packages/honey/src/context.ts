interface Page {
  name: string
  route: string
  children?: Page[]
  meta: {
    type?: string
    title?: string
    hidden?: boolean
  }
  frontMatter?: any
}

function getContext(name: string) {
  if (!(globalThis as any).__honey_internal__) {
    throw new Error(
      `Honey context not found. Please make sure you are using "${name}" of "@honeydocs/honey/context" on a Honey page.`
    )
  }
  return (globalThis as any).__honey_internal__
}

function normalizeMeta(meta: any) {
  if (typeof meta === 'string') {
    meta = {
      title: meta
    }
  }
  return meta
}

function filter(pageMap: any, activeLevel?: string) {
  let activeLevelPages: Page[] | undefined
  let meta = pageMap.find((item: any) => item.name === 'meta.json')?.meta || {}
  const metaKeys = Object.keys(meta)

  const items: Page[] = []
  for (const item of pageMap) {
    if (item.name === 'meta.json') continue
    const page: Page = {
      ...item,
      meta: normalizeMeta(meta[item.name])
    }
    if (item.children) {
      const filteredChildren = filter(item.children, activeLevel)
      page.children = filteredChildren[0]
      if (filteredChildren[1]) {
        activeLevelPages = filteredChildren[1]
      } else if (page.route === activeLevel) {
        activeLevelPages = activeLevelPages || page.children
      }
    }
    items.push(page)
  }

  return [
    items.sort((a, b) => {
      const indexA = metaKeys.indexOf(a.name)
      const indexB = metaKeys.indexOf(b.name)
      if (indexA === -1 && indexB === -1) return a.name < b.name ? -1 : 1
      if (indexA === -1) return 1
      if (indexB === -1) return -1
      return indexA - indexB
    }),
    activeLevelPages
  ]
}

export function getAllPages() {
  const internal = getContext('getAllPages')
  return filter(internal.pageMap)[0]
}

export function getCurrentLevelPages() {
  const internal = getContext('getCurrentLevelPages')
  return filter(internal.pageMap, internal.route)[1] || []
}

export function getPagesUnderRoute(route: string) {
  const internal = getContext('getPagesUnderRoute')
  return filter(internal.pageMap, route)[1] || []
}
