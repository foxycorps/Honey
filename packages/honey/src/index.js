import { HoneyPlugin, pageMapCache } from './plugin'

const defaultExtensions = ['js', 'jsx', 'ts', 'tsx']
const markdownExtensions = ['md', 'mdx']
const markdownExtensionTest = /\.mdx?$/

module.exports =
  (...args) =>
    (nextConfig = {}) => {
      const honeyConfig =
        typeof args[0] === 'string'
          ? {
            theme: args[0],
            themeConfig: args[1]
          }
          : args[0]

      const locales = nextConfig.i18n?.locales || null
      const defaultLocale = nextConfig.i18n?.defaultLocale || null

      let pageExtensions = nextConfig.pageExtensions || [...defaultExtensions]
      pageExtensions = pageExtensions.concat(markdownExtensions)

      if (locales) {
        console.log(
          '[Honey] You have Next.js i18n enabled, read here (TODO: link) for the docs.'
        )
      }

      return Object.assign({}, nextConfig, {
        pageExtensions,
        webpack(config, options) {
          const honey = new HoneyPlugin(honeyConfig)
          if (!config.plugins) {
            config.plugins = [honey]
          } else {
            config.plugins.push(honey)
          }

          config.module.rules.push({
            test: markdownExtensionTest,
            use: [
              options.defaultLoaders.babel,
              {
                loader: '@honeydocs/honey/loader',
                options: { ...honeyConfig, locales, defaultLocale, pageMapCache }
              }
            ]
          })

          if (typeof nextConfig.webpack === 'function') {
            return nextConfig.webpack(config, options)
          }

          return config
        }
      })
    }
