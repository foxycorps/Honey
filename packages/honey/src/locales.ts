import { NextRequest, NextResponse } from 'next/server'

const PUBLIC_FILE = /\.(.*)$/

export function locales(request: NextRequest) {
  const { nextUrl } = request

  const shouldHandleLocale =
    !PUBLIC_FILE.test(nextUrl.pathname) &&
    !nextUrl.pathname.includes('/api/') &&
    !nextUrl.pathname.includes('/_next/') &&
    nextUrl.locale !== ''
  if (!shouldHandleLocale) return

  // The locale code prefixed in the current URL, which can be empty.
  const fullUrl = nextUrl.toString()
  const localeInPath = fullUrl
    .slice(fullUrl.indexOf('//' + nextUrl.host) + nextUrl.host.length + 3)
    .slice(0, -(nextUrl.pathname + nextUrl.search).length)

  let finalLocale

  if (localeInPath) {
    // If a locale is explicitly set, we don't do any modifications.
    finalLocale = localeInPath
  } else {
    // If there is a locale cookie, we try to use it. If it doesn't exist or
    // it's invalid, `nextUrl.locale` will be automatically figured out by Next
    // via the `accept-languages` header.
    if (request.cookies['NEXT_LOCALE']) {
      try {
        nextUrl.locale = request.cookies['NEXT_LOCALE']
      } catch (err) {
        // The locale from the cookie isn't valid.
        // https://github.com/vercel/next.js/blob/e5dee17f776dcc79ebb269f7b7341fa6e2b6c3f1/packages/next/server/web/next-url.ts#L122-L129
      }
    }
    finalLocale = nextUrl.locale

    // Now we want to display the locale. If it's not the default one, we have
    // to prefix the URL with that locale since it's missing. Only the default
    // locale can be missing from there for consistency.
    if (finalLocale !== nextUrl.defaultLocale) {
      return NextResponse.redirect(
        new URL(
          '/' + finalLocale + nextUrl.pathname + nextUrl.search,
          request.url
        )
      )
    }
  }

  let pathname = nextUrl.pathname || '/'
  if (pathname === '/') pathname += 'index'

  // If we are not showing the correct localed page, rewrite the current request.
  if (!pathname.endsWith('.' + finalLocale)) {
    return NextResponse.rewrite(
      new URL(
        '/' + finalLocale + pathname + '.' + finalLocale + nextUrl.search,
        request.url
      )
    )
  }
}

export function withLocales(middleware: any) {
  return (...args: any[]) => {
    return locales(args[0]) || middleware(...args)
  }
}
