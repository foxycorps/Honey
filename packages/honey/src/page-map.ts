import { PageMapItem } from './types'
import { getLocaleFromFilename, existsSync } from './utils'
import path from 'path'
import filterRouteLocale from './filter-route-locale'
export const extension = /\.mdx?$/
export const metaExtension = /meta\.?([a-zA-Z-]+)?\.json/

export function findPagesDir(dir: string = process.cwd()): string {
  // prioritize ./pages over ./src/pages
  if (existsSync(path.join(dir, 'pages'))) return 'pages'
  if (existsSync(path.join(dir, 'src/pages'))) return 'src/pages'
  if (existsSync(path.join(dir, 'docs'))) return 'docs'
  if (existsSync(path.join(dir, 'src/docs'))) return 'src/docs'

  throw new Error(
    "> Couldn't find a `pages` directory. Please create one under the project root"
  )
}

export function getPageMap(
  currentResourcePath: string,
  pageMaps: PageMapItem[],
  fileMap: Record<string, PageMapItem>,
  defaultLocale: string
) {
  const activeRouteLocale = getLocaleFromFilename(currentResourcePath)
  const pageItem = fileMap[currentResourcePath]
  const metaPath = path.dirname(currentResourcePath)
  const metaExtension = activeRouteLocale ? `${activeRouteLocale}.json` : `json`
  const pageMeta =
    fileMap[`${metaPath}/meta.${metaExtension}`]?.meta?.[pageItem.name]
  const title =
    (typeof pageMeta === 'string' ? pageMeta : pageMeta?.title) || pageItem.name
  if (activeRouteLocale) {
    return [
      filterRouteLocale(pageMaps, activeRouteLocale, defaultLocale),
      fileMap[currentResourcePath].route,
      title
    ]
  }
  return [pageMaps, fileMap[currentResourcePath].route, title]
}
