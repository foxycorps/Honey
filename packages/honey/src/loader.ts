import type { LoaderOptions } from './types'

import path from 'path'
import grayMatter from 'gray-matter'
import slash from 'slash'
import { LoaderContext } from 'webpack'
import { Repository } from '@napi-rs/simple-git'

import { addPage } from './content-dump'
import { getLocaleFromFilename } from './utils'
import { compileMdx } from './compile'
import { getPageMap, findPagesDir } from './page-map'
import { collectFiles } from './plugin'
const extension = /\.mdx?$/
const isProductionBuild = process.env.NODE_ENV === 'production'

// TODO: create this as a webpack plugin.
const indexContentEmitted = new Set()

const pagesDir = path.resolve(findPagesDir())

let [repository, gitRoot] = (function () {
  try {
    const repo = Repository.discover(process.cwd())
    // repository.path() returns the `/path/to/repo/.git`, we need the parent directory of it
    const gitRoot = path.join(repo.path(), '..')
    return [repo, gitRoot]
  } catch (e) {
    console.warn('Init git repository failed', e)
    return []
  }
})()

async function loader(
  context: LoaderContext<LoaderOptions>,
  source: string
): Promise<string | Buffer> {
  context.cacheable(true)

  const options = context.getOptions()
  let {
    theme,
    themeConfig,
    defaultLocale,
    unstable_flexsearch,
    unstable_staticImage,
    mdxOptions,
    pageMapCache
  } = options

  const { resourcePath } = context
  const filename = resourcePath.slice(resourcePath.lastIndexOf('/') + 1)
  const fileLocale = getLocaleFromFilename(filename)

  // Check if there's a theme provided
  if (!theme) {
    throw new Error('No Honey theme found!')
  }

  let pageMapResult, fileMap
  if (isProductionBuild) {
    const data = pageMapCache.get()!
    pageMapResult = data.items
    fileMap = data.fileMap
  } else {
    const data = await collectFiles(pagesDir, '/')
    pageMapResult = data.items
    fileMap = data.fileMap
  }
  const [pageMap, route, title] = getPageMap(
    resourcePath,
    pageMapResult,
    fileMap,
    defaultLocale
  )

  if (!isProductionBuild) {
    // Add the entire directory `pages` as the dependency
    // so we can generate the correct page map.
    context.addContextDependency(pagesDir)
  } else {
    // We only add meta files as dependencies for prodution build,
    // so we can do incremental builds.
    Object.entries(fileMap).forEach(([filePath, { name, meta, locale }]) => {
      if (
        name === 'meta.json' &&
        meta &&
        (!fileLocale || locale === fileLocale)
      ) {
        context.addDependency(filePath)
      }
    })
  }

  // Extract frontMatter information if it exists
  let { data, content } = grayMatter(source)

  let layout = theme
  let layoutConfig = themeConfig || null

  // Relative path instead of a package name
  if (theme.startsWith('.') || theme.startsWith('/')) {
    layout = path.resolve(theme)
  }
  if (layoutConfig) {
    layoutConfig = slash(path.resolve(layoutConfig))
  }

  if (isProductionBuild && indexContentEmitted.has(filename)) {
    unstable_flexsearch = false
  }

  const { result, titleText, headings, hasH1, structurizedData } =
    await compileMdx(
      content,
      mdxOptions,
      {
        unstable_staticImage,
        unstable_flexsearch
      },
      resourcePath
    )
  content = result
  content = content.replace('export default MDXContent;', '')

  if (unstable_flexsearch) {
    // We only add .MD and .MDX contents
    if (extension.test(filename) && data.searchable !== false) {
      await addPage({
        fileLocale: fileLocale || 'default',
        route,
        title,
        data,
        structurizedData
      })
    }

    indexContentEmitted.add(filename)
  }

  let timestamp: number | undefined
  if (repository && gitRoot) {
    if (repository.isShallow()) {
      if (process.env.VERCEL) {
        console.warn(
          `The repository is shallow cloned, so the latest modified time will not be presented. Set the VERCEL_DEEP_CLONE=true environment variable to enable deep cloning.`
        )
      } else if (process.env.GITHUB_ACTION) {
        console.warn(
          `The repository is shallow cloned, so the latest modified time will not be presented. See https://github.com/actions/checkout#fetch-all-history-for-all-tags-and-branches to fetch all the history.`
        )
      } else {
        console.warn(
          `The repository is shallow cloned, so the latest modified time will not be presented.`
        )
      }
    }
    try {
      timestamp = await repository.getFileLatestModifiedDateAsync(
        path.relative(gitRoot, resourcePath)
      )
    } catch (e) {
      // Failed to get timestamp for this file. Silently ignore it.
    }
  }

  const prefix =
    `import __honey_withLayout__ from '${layout}'\n` +
    `import { withSSG as __honey_withSSG__ } from '@honeydocs/honey/ssg'\n` +
    `${layoutConfig
      ? `import __honey_layoutConfig__ from '${layoutConfig}'`
      : ''
    }\n\n` +
    `const __honey_pageMap__ = ${JSON.stringify(pageMap)}\n` +
    `globalThis.__honey_internal__ = {\n` +
    `  pageMap: __honey_pageMap__,\n` +
    `  route: ${JSON.stringify(route)},\n` +
    `}\n` +
    `
    const __honey_content__ = <MDXContent/>
    const HoneyLayout = __honey_withSSG__(__honey_withLayout__({
      filename: "${slash(filename)}",
      route: "${slash(route)}",
      meta: ${JSON.stringify(data)},
      pageMap: __honey_pageMap__,
      titleText: ${JSON.stringify(titleText)},
      headings: ${JSON.stringify(headings)},
      hasH1: ${JSON.stringify(hasH1)},
      ${timestamp ? `timestamp: ${timestamp},\n` : ''}
    }, ${layoutConfig ? '__honey_layoutConfig__' : 'null'}))
    `

  const suffix = `export default function HoneyPage (props) {
  return <HoneyLayout {...props}>{__honey_content__}</HoneyLayout>
}
HoneyPage.getLayout = HoneyLayout.getLayout`

  // console.log(content)

  // Add imports and exports to the source
  return prefix + '\n\n' + content + '\n\n' + suffix
}

export default function syncLoader(
  this: LoaderContext<LoaderOptions>,
  source: string,
  callback: (err?: null | Error, content?: string | Buffer) => void
) {
  loader(this, source)
    .then(result => callback(null, result))
    .catch(err => callback(err))
}
