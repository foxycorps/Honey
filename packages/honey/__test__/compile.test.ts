import { compileMdx } from '../src/compile'
import { it, describe, expect } from 'vitest'
import path from 'path'
import fs from 'fs/promises'

function loadFixture(name: string) {
  const filePath = path.join(
    process.cwd(),
    '__test__',
    'fixture',
    'headings',
    name
  )
  return fs.readFile(filePath, 'utf8')
}

describe('process heading', () => {
  it('static-h1', async () => {
    const data = await loadFixture('static-h1.mdx')
    const result = await compileMdx(data)
    expect(result).toMatchSnapshot()
  })
  it('dynamic-h1', async () => {
    const data = await loadFixture('dynamic-h1.mdx')
    const result = await compileMdx(data)
    expect(result).toMatchSnapshot()
  })
  it('no-h1', async () => {
    const data = await loadFixture('no-h1.mdx')
    const result = await compileMdx(data)
    expect(result).toMatchSnapshot()
  })
})
