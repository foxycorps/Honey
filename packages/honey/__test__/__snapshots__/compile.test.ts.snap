// Vitest Snapshot v1

exports[`process heading > dynamic-h1 1`] = `
{
  "hasH1": true,
  "headings": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 22,
              "line": 8,
              "offset": 153,
            },
            "start": {
              "column": 3,
              "line": 8,
              "offset": 134,
            },
          },
          "type": "text",
          "value": "Posts Tagged with “",
        },
        {
          "attributes": [],
          "children": [],
          "data": {
            "_mdxExplicitJsx": true,
          },
          "name": "TagName",
          "position": {
            "end": {
              "column": 32,
              "line": 8,
              "offset": 163,
            },
            "start": {
              "column": 22,
              "line": 8,
              "offset": 153,
            },
          },
          "type": "mdxJsxTextElement",
        },
        {
          "position": {
            "end": {
              "column": 33,
              "line": 8,
              "offset": 164,
            },
            "start": {
              "column": 32,
              "line": 8,
              "offset": 163,
            },
          },
          "type": "text",
          "value": "”",
        },
      ],
      "depth": 1,
      "position": {
        "end": {
          "column": 33,
          "line": 8,
          "offset": 164,
        },
        "start": {
          "column": 1,
          "line": 8,
          "offset": 132,
        },
      },
      "type": "heading",
      "value": "Posts Tagged with “”",
    },
  ],
  "result": "/*@jsxRuntime automatic @jsxImportSource react*/
import {useMDXComponents as _provideComponents} from \\"@mdx-js/react\\";
import {useRouter} from 'next/router';
export const TagName = () => {
  const {tag} = useRouter().query;
  return tag || null;
};
function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = Object.assign({}, _provideComponents(), props.components);
  return MDXLayout ? <MDXLayout {...props}><_createMdxContent /></MDXLayout> : _createMdxContent();
  function _createMdxContent() {
    const _components = Object.assign({
      h1: \\"h1\\"
    }, _provideComponents(), props.components);
    return <_components.h1>{\\"Posts Tagged with “\\"}<TagName />{\\"”\\"}</_components.h1>;
  }
}
export default MDXContent;
",
  "structurizedData": {},
}
`;

exports[`process heading > no-h1 1`] = `
{
  "hasH1": false,
  "headings": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 6,
              "line": 1,
              "offset": 5,
            },
            "start": {
              "column": 4,
              "line": 1,
              "offset": 3,
            },
          },
          "type": "text",
          "value": "H2",
        },
      ],
      "depth": 2,
      "position": {
        "end": {
          "column": 6,
          "line": 1,
          "offset": 5,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "heading",
      "value": "H2",
    },
  ],
  "result": "/*@jsxRuntime automatic @jsxImportSource react*/
import {useMDXComponents as _provideComponents} from \\"@mdx-js/react\\";
function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = Object.assign({}, _provideComponents(), props.components);
  return MDXLayout ? <MDXLayout {...props}><_createMdxContent /></MDXLayout> : _createMdxContent();
  function _createMdxContent() {
    const _components = Object.assign({
      h2: \\"h2\\"
    }, _provideComponents(), props.components);
    return <_components.h2>{\\"H2\\"}</_components.h2>;
  }
}
export default MDXContent;
",
  "structurizedData": {},
}
`;

exports[`process heading > static-h1 1`] = `
{
  "hasH1": true,
  "headings": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 14,
              "line": 1,
              "offset": 13,
            },
            "start": {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "type": "text",
          "value": "Hello World",
        },
      ],
      "depth": 1,
      "position": {
        "end": {
          "column": 14,
          "line": 1,
          "offset": 13,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "heading",
      "value": "Hello World",
    },
  ],
  "result": "/*@jsxRuntime automatic @jsxImportSource react*/
import {useMDXComponents as _provideComponents} from \\"@mdx-js/react\\";
function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = Object.assign({}, _provideComponents(), props.components);
  return MDXLayout ? <MDXLayout {...props}><_createMdxContent /></MDXLayout> : _createMdxContent();
  function _createMdxContent() {
    const _components = Object.assign({
      h1: \\"h1\\"
    }, _provideComponents(), props.components);
    return <_components.h1>{\\"Hello World\\"}</_components.h1>;
  }
}
export default MDXContent;
",
  "structurizedData": {},
  "titleText": "Hello World",
}
`;
